richtextfx/src/main/java/org/fxmisc/richtext/model/SegmentOps.java

```diff
@@ -0,0 +1,42 @@
+package org.fxmisc.richtext.model;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+
+
+/**
+ * Defines the operations which are supported on a specific segment type.
+ *
+ * @param <SEG>
+ * @param <S>
+ */
+public interface SegmentOps<SEG, S> {
+    public int length(SEG seg);
+
+    public char charAt(SEG seg, int index);
+
+    public String getText(SEG seg);
+
+    public SEG subSequence(SEG seg, int start, int end);
+
+    public SEG subSequence(SEG seg, int start);
+
+    public SEG append(SEG seg, String str);
```

Comment:
Meaning, if SegmentOps was only:
````java
public interface SegmentOps<SEG, S> {
    char charAt(SEG seg, int index);
    int length(SEG seg);
    String getText(SEG seg);
    boolean canJoin(SEG currentSeg, SEG nextSeg);
    S getStyle(SEG seg);
    SEG setStyle(SEG seg);
}
````
then `TextOps` could be
````java
public interface TextOps<SEG, S> extends SegmentOps<SEG, S> {
    SEG subsequence(SEG seg, int start, int end);
    SEG append(SEG seg, String text);
    // other text-related methods
}
````
and then `StyledText` could implement `TextOps` as well as other classes like `Hyperlink` or some other custom class that is text-based.

.... But, having written all of that, I don't see why we would do that since `Hyperlink` would have a separate model (`displayedText` and `linkText`) and use a separate `SegmentOps` implementation anyway... :sweat_smile: :laughing: